### ISSUE TITLE
Suggestion: tactical usage of prefix tries + charCodeAt()

### ISSUE THREAD

**Issue by JobLeonard** [bvaughn/js-search#42]
So [I just spent a week or so on optimising lz-string](https://github.com/pieroxy/lz-string/pull/98). I'm not suggesting you read through all of that, because a lot of stuff is going on there and most of it is irrelevant to `js-search`. However, the biggest improvement came from changing data structures, and I think a similar could work for (parts of) `js-search` too.

Ignoring the details of the LZ compression algorithm, the way `lz-string` worked before was to build a dictionary of previously encountered string segments. In simplified terms it did so by using `charAt()` and string concatenation, and storing these string segments as keys in a JS object used as the dictionary.

My alternative approach is to use [`charCodeAt()`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/String/charCodeAt) instead to build a [prefix trie](https://en.wikipedia.org/wiki/Trie) out of integer values. This is faster, sometimes _much_ faster in fact. There are at least two reasons for that:

- most modern JS engines have optimisations for objects that only have integer keys, so when used properly the constant overhead can be smaller
- it actually knocks (at least part of) an _n_ of it's Big O performance, because string concatenation and hashing depends on string length (how much is not entirely clear [since some browsers use ropes internally](https://blog.mozilla.org/javascript/2014/07/21/slimmer-and-faster-javascript-strings-in-firefox/)), whereas adding new nodes with integer keys is constant (ignoring the situation where the bucket needs to expand).

That last bit matters _a lot_ - strings that are really long and compress really well compressed an _order of magnitude faster_ on mobile browsers and Firefox.

Now, I'm well aware that implementing this would basically mean rethinking and rewriting the entire library, and make the code pretty complicated for people not familiar with this type of data structure, but perhaps it could be applied selectively in some places where it really is worth it.

One potential big win that I see would be using this for the indexing. If we assume that string concatenation has O(n) overhead, then [PrefixIndexStrategy](https://github.com/bvaughn/js-search/blob/master/source/IndexStrategy/PrefixIndexStrategy.js) is O(n²), and [AllSubstringsIndexStrategy](https://github.com/bvaughn/js-search/blob/master/source/IndexStrategy/AllSubstringsIndexStrategy.js) is _**O(n³)**_! Reducing that to O(n) and O(n²) performance would be a big win for larger inputs.

This is not entirely hypothetical: I'm working with gene data sets of tens of thousands of entries. Luckily we only search by prefix, but before I implemented that the indexing took up tens of seconds. Our datasets are growing however, so I'm kind of worried about the future...

(aside: as an example of what probably _isn't_ worth it: instead of the current tokenizing strategy, you could build a prefix trie by doing a single pass across the whole string and "simply" treat all charCodes that represent white-space as the end of a prefix. That would save the overhead of creating a new array of strings, but not improve Big O performance. It would make it hard to do more complicated tokenization so it's probably not worth it)

Anyway, just writing this out took an hour and I have other work to do (if I had a month to spare I would give this a try myself), but I just wanted to bring this possibility to your attention.

**Comment by bvaughn** [comment]
Thanks for the fantastic write-up @JobLeonard! Much appreciated!

I have a lot of projects competing for limited time at the moment but I would love to make some time at some point to dig into this further.

**Comment by JobLeonard** [comment]
I know that feeling all too well. We'll see which one of us will be first to have enough time to spare to try this out, if ever ;)

**Comment by JobLeonard** [comment]
I figured I'd at least write out a simple code example of `charCodeAt()`-based tries:
```js
function createPrefixTrie(list) {
  let root = {
    0: []
  };
  for (let i = 0; i < list.length; i++) {
    let word = list[i];
    let node = root;
    for (let j = 0; j < word.length; j++) {
      // offset by one, because node[0] is where
      // we store the indices of list items matching
      // the current prefix
      let char = word.charCodeAt(j) + 1;
      if (node[char]) {
        // prefix already exists in trie
        node = node[char];
        // store index of current word at this prefix
        node[0].push(i);
      } else {
        let newPrefix = {
          0: [i]
        };
        node[char] = newPrefix;
        node = newPrefix;
      }
    }
  }
  return root;
}

function createSubstringTrie(list) {
  let root = {
    0: []
  };
  for (let i = 0; i < list.length; i++) {
    let word = list[i];
    let nodes = [];
    for (let j = 0; j < word.length; j++) {
      // we're going to find all substrings in this word
      // by starting again from the root for each charCode
      // while continuing with the previous substrings.
      nodes.push(root);
      // offset by one, because node[0] is where
      // we store the indices of list items matching
      // the current substring
      let char = word.charCodeAt(j) + 1;
      for (let k = 0; k < nodes.length; k++) {
        let node = nodes[k];
        let nextNode = node[char];
        if (nextNode) {
          // substring already exists in trie
          nodes[k] = nextNode;
          // substrings can repeat, so avoid
          // adding them multiple times.
          if (nextNode[0].indexOf(i) == -1) {
            // store index of current word at this substring
            nextNode[0].push(i);
          }
        } else {
          nextNode = {
            0: [i]
          };
          node[char] = nextNode;
          nodes[k] = nextNode;
        }
      }
    }
  }
  return root;
}

function findMatches(str, trie, list) {
  for (let i = 0; i < str.length; i++) {
    let char = str.charCodeAt(i) + 1;
    if (trie[char]) {
      trie = trie[char];
    } else {
      // prefix does not match anything in the trie
      return [];
    }
  }
  let indices = trie[0];
  return indices.map((idx) => {
    return list[idx];
  });
}

const testList = ['foo', 'bar', 'baz'];
const testPrefixTrie = createPrefixTrie(testList);
console.log('prefix trie created from ["foo", "bar", "baz"]: ', testPrefixTrie);

const testSubstringTrie = createSubstringTrie(testList);
console.log('substring trie created from ["foo", "bar", "baz"]: ', testSubstringTrie);

const prefixMatches = findMatches('ba', testPrefixTrie, testList);
console.log('matches for "ba" prefix: ', prefixMatches);

const substringMatches1 = findMatches('a', testSubstringTrie, testList);
console.log('matches for "a" substring: ', substringMatches1);

const substringMatches2 = findMatches('ar', testSubstringTrie, testList);
console.log('matches for "ar" substring: ', substringMatches2);
```
https://jsfiddle.net/vanderZwan/pbq3oe8g/1/

I hope this simplifies future work :)

**Comment by bvaughn** [comment]
That's fantastic 😁 Thanks for the jump-start!

**Comment by JobLeonard** [comment]
Cheers! I wanted to write it all down before I forgot again :)

Note that this version of substring-trie might have a problem: if my original list is `['foo', 'ook']` and I search for `'oo'`, it will return `['foo', 'ook']`. However, the desired behaviour is more likely to be `['ook', 'foo']` - adding a bit of logic to rank by how early in the string substring occurred shouldn't be that difficult though.

**Comment by bvaughn** [comment]
Started experimenting with this today. Not sure yet how it's going to affect the API (probably will require a new major release) but I'm pretty excited by the perf improvements! Based on my own benchmark, creating an index is 2-5 times faster depending on the browser I'm testing. 👍

**Comment by bvaughn** [comment]
I've argued before that js-search is (a bit) slower than lunr to build an index (something that happens once, and can maybe even be done beforehand) but faster (by a good bit) to actually run a search (something that happens many times, in response to user feedback) and so the latter was a more important performance metric.

The new approach is slower in Chrome, faster in Firefox, and about the same in Safari (see [benchmark](https://jsperf.com/js-search-searching-index)).

**Comment by JobLeonard** [comment]
I was actually thinking about this the other day: for search you would need to walk through the whole prefix within the trie, whereas for a dictionary it's a string-based lookup. The latter can use whatever internal datastructure is used and avoid the JS object lookups, so that can give it an edge.

One potential work-around could be a flattening step in the indexing: after building the trie, convert it to the old flat string-based object. This _might_ still be faster because you would avoid repeated string creation; every prefix string would only be created once, during flattening. Of course, this is likely to be slower on Firefox and Safari.

Another question is whether or not it is actually true that the search is the more important performance metric: yes, if the search was in the order of seconds or even tens of seconds, that is jank you want to avoid. But on my machine the trie approach has 2,287,180 ops per second, and the old 3,761,752 ops per second, searching eleven prefixes. That's about four, five orders of magnitude faster than my best WPM scores :P

BTW, I think I see a few issues with your benchmark:

- function creation is part of the code being benchmarked in the  `js-search` benchmark. In reality, this would have already been compiled

- this is a very tiny list. for a more realistic benchmark on both ends I'd argue we need much larger input - the amount of options we have in our usage of react-virtualised-select` is between 1000 and 10000 for example , for example). [Here's a github repo with 479k words](https://github.com/dwyl/english-words)

I made another benchmark that takes this into account (on jsbench.github.io, because jsperf wouldn't let me fork your code for some reason). I used the first 10,000 words from the linked repo. I also shuffle it before indexing (so they're not in alphabetical order), and then choose 100 random words to search for. This shuffling/random word search happens every run, so we should get a bit more of a realistic random distribution of inputs.

[indexing](http://jsbench.github.io/#8f65134bd0fd145a95a96b60d2057434)

[search](http://jsbench.github.io/#9b70178989d5bd6c77a3245629c5d0ae)

In this benchmark the new way of searching is stll slower on Chrome: 262k vs 147k searches per second. Wait, no, multiply that by 100, since I search 100 different needles ;). So I don't think the search is going to be the bottleneck. 

On the other hand, creating an index is in the order of _tens_ of operations per second. If the input gets much larger (and this is not unrealistic for me), that becomes dang slow.

Sadly, the new way of indexing seems slower :(. I'm a little bit confused why though - I would expect that with larger datasets, the trie would perform better (for the same reason it works better with lz-string). The only big difference that I can think of is that with lz-string, the trie stores a number when it's created and then leaves it alone. Here it stores an array that keeps growing. But I don't really see why that would make it lose its advantage over the flat dictionary.

**Comment by bvaughn** [comment]
Good point about the creation of the function being in the test case. However moving it out didn't seem to have much of an impact. (Just tried it and verified that the overall results seem the same as before.) Also fair point about the size and variety of the list.

I've adjusted both of my benchmarks to use the 10k word list you referenced ([create index](https://jsperf.com/js-search-index-strategies), [search index](https://jsperf.com/js-search-searching-index/)). It looks like, with the larger data set, the character code / trie approach is faster for both indexing _and_ searching for all browsers except Safari, which is a _tiny_ bit slower to search.

**Comment by bvaughn** [comment]
Another weird data point. I decided to try the trie approach in another, simpler search lib of mine, [`js-worker-search`](https://github.com/bvaughn/js-worker-search). Unfortunately the version I tested with the trie in place ([ddfcdfd](https://github.com/bvaughn/js-worker-search/commit/ddfcdfd0e7b3748f6371d01c6f61376e98e7a4c0)) runs way slower than the version with the letter-map ([75a1281](https://github.com/bvaughn/js-worker-search/commit/75a1281bdd55d64521d447314584c7fdf346e22e)) based on my own, potentially flawed [benchmarking](https://github.com/bvaughn/js-worker-search/tree/master/benchmarks).

**Comment by JobLeonard** [comment]
I am getting the same results as you are on jsperf, but that contradicts the results on jsbench (which favoured: Chrome is faster in the former, slower in the latter . Which is confusing because as far as I can see this is the same code.

Having said that, I think I trust jsperf a bit better since it's a generated PHP site instead of a dynamically created one. That probably means less risk of deopting and other JIT-weirdness going on.

```js
type Node = {
  [charCode: any]: Node,
  "0": Array<any>
};
```

Eh... that should be a number, not a string. Could the problem be that you're effectively telling your compiler to add string conversion code? Because that would undo the whole speed advantage...

**Comment by bvaughn** [comment]
Hm. Flow doesn't support that, at least not with the syntax I was using, but you may be onto something there. I'll try an alternate approach.

My (limited) understanding of object properties is that, according to ES spec, they're converted to strings anyway. I thought maybe the perf improvements you were seeing came from the char-code-range or length of the keys or something... I'm not sure.

I'll take another pass at it. 😄

**Comment by bvaughn** [comment]
Re-built the trie implementation with some modifications to make Flow happy:
https://github.com/bvaughn/js-worker-search/compare/poc-for-job?expand=1

Test results are still meh:
<img width="478" alt="screen shot 2017-07-16 at 6 28 36 pm" src="https://user-images.githubusercontent.com/29597/28253157-cff4633a-6a54-11e7-9382-0363d089f887.png">

Am I missing something? ☹️

**Comment by JobLeonard** [comment]
> My (limited) understanding of object properties is that, according to ES spec, they're converted to strings anyway. I thought maybe the perf improvements you were seeing came from the char-code-range or length of the keys or something... I'm not sure.

They only have to _behave_ that way, which leaves the door open for internal optimisations. [See this SO answer](https://stackoverflow.com/a/10639526/3211791) with [this JSperf benchmark to prove it](https://jsperf.com/string-vs-integer-object-indices). What you can already see there however is that the biggest benefit is look-up speed.

>  Test results are still meh

Yeah, I think I have some bad news :(

Note: I haven't looked at search performance yet, I really should get back to my actual work ;)

I looked at your [benchmark code for creating indices](https://jsperf.com/js-search-index-strategies) again and found the reason why we had different results for index creation: **it didn't re-initialise the index every time**. So what ended up happening is basically a bunch of indexes being pushed to _an already created array in an already created object_, whereas what should be happening is that both object and array would be created fresh on each run. The fact that the results are so different once the dictionary/trie is initialised tells us something about what we're measuring and what impact it has.

So let's try to reason about what happens under the hood.

If we don't re-initialise the dictionary/trie, we don't create any new objects in the trie after a while. So that leaves lookup speed, and the speed at which arrays are expanded (because we keep pushing to the same array). The latter might vary per implementation, but as I understand the rule of thumb for vector implementations is that the backing array doubles in size when we fill a vector up to capacity. If that actually would go from 1, 2, 4, 8, etc, that would mean the backing array would have to expand quite often in the beginning, but after a while  it wouldn't, because exponential growth (256, 512, 1024).

So, with that in mind, if we benchmark this often enough, eventually doubling the backing array makes the capacity so much larger that it doesn't have to expand for any more runs. At that point pushing to the array is just writing to a pre-allocated array. In that case the `charCode` trie seems to win, because at that point look-up becomes the bottleneck, and it looks like `charCode` based lookup of objects is faster.

This would also explain why my results with `lz-string` are so different: there, we are creating a symbol table, and so we don't push to an array of indices every time we pass a node. So the growth of the backing arrays seem to be the bottlenecks here in this (flawed) benchmark.

But of course, this is not representative of what is happening when we initiate an index from scratch!

I changed the code to fix that, added a string-based trie version to compare the two. I also added the word shuffling back, to reduce the chance that there is a systemic bias introduced by the word order (it doesn't seem to affect anything, however).

https://jsperf.com/index-strategies/1

Sadly, we're back to my previous findings: `charCode` is slower in Chrome, faster on Firefox. The Map versions are faster, but Firefox likes the `char` version more, and Chrome likes the `charCode` version more:

|                        | Chrome        | Firefox        |
|------------------------|---------------|----------------|
| `char`                 | 21.72 ± 4.01% | 28.87 ± 5.92%  |
| `charCode` trie        | 14.71 ± 6.53% | 33.01 ± 8.71%  |
| `char` (Map)           | 32.59 ± 2.68% | 34.22 ± 5.59%  |
| `charCode` trie (Map)  | 64.05 ± 3.70% | 29.74 ± 17.97% |
| `charCode`-string trie | 14.60 ± 5.72% | 28.03 ± 8.79%  |

That suggests that for trie creation, object and array creation/expansion are the bottlenecks (this actually fits my findings with `lz-string` very well).

So in short: AAAAARGH! Benchmarking is hard! Results are complicated! And this optimisation doesn't seem to work in this scenario :(

But we had fun getting an idea about what JS engines do under the hood I guess :P

Anyway, I guess we should let this be for now, until Maps are all around us (and hopefully, Firefox will implement optimisations for integer keys). Or until we come up with a better data structure than tries with arrays, but I'm no Donald Knuth so  ¯\\\_(ツ)_/¯ 

(Although this would suggest that for server-side Node.js code, which is guaranteed to be the latest V8 engine and probably deals with much larger data, charCode Maps are the best option. Well, assuming search is faster too, or at least not worse)

**Comment by JobLeonard** [comment]
So I couldn't help but think about this over lunch... TL;DR: none of the things below help (although limiting the number of stored indices does give a small speed-up, which might be useful in specific scenarios).

I'm just writing this down to others don't have to try and fail with the same approach.

The overal idea was that if object creation and memory allocation are the main bottlenecks, then perhaps pre-initialising arrays and limiting the number of stored indices will help.

[I updated my benchmarks with the ideas from below](https://jsperf.com/index-strategies
)

# Pre-initialisation

In JS, you can actually assign a number to `<some array>.length` to force its length, but depending on the implementation that can leave the backing array untouched. So what might work is something like this:

```js
    if (!Array.isArray(charObject[prefix])) {
      // try to force backing array to initialise as dense, 256-entry      
      var newIdx = [i,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0];
      newIdx.length = 1;
      charObject[prefix] = newIdx;
```

I added tests for all four options with pre-initiated arrays... it's slower for all of them. Thinking about this some  more, this makes sense: pre-initialising only makes sense for the first few letters in a prefix, since we'll hone in on a single word after only a few letters. Let's say I want to see how many characters I need to identify "abovestairs". In our dataset, the first letter of a word is _always_ "a", so it will match all 10k words. However, "ab" is already down to 1178 matches, "abo" down to 134, "abov" and "above" both have 9 matches, and "aboves" is the last prefix with more than one match.

So it makes sense that this is slower: even if this were advantageous for short prefixes, for the leaves (which make up the majority of nodes in the trie) this is a step back.

# Limiting the number of stored indices

This obviously won't always be viable, but if we only care about, say, the first three or four matching prefixes (like with an autocomplete), limiting the array expansion might do the trick.

```js
    var indices = charObject[prefix]
    if (!Array.isArray(indices)) {
      charObject[prefix] = [i];
    } else if (indices.length < 4) {
      indices.push(i);
    }
```

Answer: yes, it's faster, for all indexing strategies, but not by a lot (again, probably because it only takes a handful of characters before we're almost guaranteed to never have more than a handful matches). It does show that expanding the array is part of the slowdown, however.

# Construction in two phases

I came up with this alternate way of construction, thinking it had a very slim chance of being faster: doing this in two phases. First we only count _how many_ times a prefix occurs, then we initiate arrays at exactly the right size (as a bonus, we can now use typed arrays), then we will said arrays.

However, it turned out this was slower for all four indexing strategies. Presumably, looking up and writing to the object, even if that value is only an integer, is a heavy enough operation that we want to avoid it.

It _might_ actually perform better on the search side though, since typed arrays generally perform better than arrays.

Here is the example code, for posterity:

```js
var charCodeObject = {}, nodes = [];
// generate trie and count how many indices we will store
for (var i = 0; i < words.length; i++) {
  var word = words[i];
  var node = charCodeObject;
  for (var j = 0; j < word.length; j++) {
    var char = word.charCodeAt(j) + 1;
    var newPrefix = node[char];
    if (newPrefix) {
      newPrefix[0]++;
    } else {
      newPrefix = {
        0: 1,
      };
      nodes.push(newPrefix);
      node[char] = newPrefix;
    }
    node = newPrefix;
  }
}

// Float64 can store all integers up to 54 bit.
// If you need to index more than 1000^5 words,
// I don't think you should be using JavaScript.
var ArrayConstr = words.length < 256 ? Uint8Array : words.length < 65536 ? Uint16Array : words.length < 4294967296 ? Uint32Array : Float64Array;
for (var i = 0; i < nodes.length; i++){
  var node = nodes[i];
  var l = node[0]
  var indices = new ArrayConstr(l);
  node[0] = indices;
}

// Walk through the trie again, now filling up the indices
// Note that we now are guaranteed that all nodes exist,
// and that all arrays are pre-allocated
for (var i = 0; i < words.length; i++) {
  var word = words[i];
  var node = charCodeObject;
  for (var j = 0; j < word.length; j++) {
    var char = word.charCodeAt(j) + 1;
    var newPrefix = node[char];
    var indices = newPrefix[0];
    // we keep track of where in the indices array we are
    // in the *last* element of the array. When we reach
    // that element, it overwrites itself.
    // We can safely do this because we're guaranteed that
    // the array will fill up exactly to its length
    // equivalent to:
    // var idx = indices[indices.length-1]++; // postfix returns old value
    // indices[idx] = i;
    indices[indices[indices.length-1]++] = i;
    node = newPrefix;
  }
}
```

**Comment by JobLeonard** [comment]
On the way home I came up with another option: reverse the array and the map. That is, store the dictionary in index 0 of the indices array, and push the indices to that array:

```js
var charCodeInv = [{}];

for (var i = 0; i < words.length; i++) {
  var word = words[i];
  var node = charCodeInv;
  for (var j = 0; j < word.length; j++) {
    var char = word.charCodeAt(j);
    var newPrefix = node[0][char];
    if (newPrefix) {
      newPrefix.push(i);
    } else {
      newPrefix = [{}, i];
      node[char] = newPrefix;
    }
    node = newPrefix;
  }
}

// Map version
var charCodeMap = [new Map()];
for (var i = 0; i < words.length; i++) {
  var word = words[i];
  var map = charCodeMap;
  for (var j = 0; j < word.length; j++) {
    var char = word.charCodeAt(j);
    var prefixMap = map[0].get(char);
    if (prefixMap) {
      prefixMap.push(i);
    } else {
      prefixMap = [new Map(), i];
      map[0].set(char, prefixMap);
    }
    map = prefixMap;
  }
}
```

The benefit would be that constructing the trie would require two array lookups and one dictionary lookup instead of two dictionary lookups and one array lookup. The downside might be that the array no longer is just integers, which some JS engines supposedly JIT-optimise.

Another downside is that walking through the trie while searching now requires an array lookup and a dictionary lookup for each character, whereas before it was just a dictionary lookup. So that's probably slower...

Anyway, indexing benchmarks:

https://jsperf.com/trie-vs-inverted-trie

|                           | Chrome        | Firefox        |
|---------------------------|---------------|----------------|
| `charCode` trie           | 14.46 ± 8.98% | 35.18 ± 5.91%  |
| inverted `charCode` trie  | 23.56 ± 1.19% | 49.03 ± 2.90%  |
| `charCode` map            | 67.47 ± 3.65% | 29.27 ± 18.75% |
| inverted `charCode` Map   | 53.57 ± 5.49% | 35.46 ± 19.36% |
| char-string dictionary    | 21.49 ± 3.57% | 29.85 ± 5.56%  |
| char-string map           | 33.41 ± 2.62% | 35.25 ± 3.94%  |

(ops/sec, higher is better)

Oh well, it was worth a shot.

**Comment by bvaughn** [comment]
Wow 😁  So much info! Thanks for the research and for taking time to write it all down and share it!

I don't have time to respond in depth now (work, Monday morning) but I'll revisit this at some point and play with it some more. Maybe next weekend...

**Comment by JobLeonard** [comment]
I think I've got it:

- typically, we'll be dealing with a small alphabet (Latin alphabet is 26 letters, times two for capitalisation, plus punctuation and some special characters)
- of said alphabet, we are probably only using a subset in our words
- of said subset, _as soon as we're beyond the root_, we'll have an even smaller subset

With that in mind, a linear search through an array is probably going to be faster than any fancy algorithm, the same way that insertion sort beats everything else for arrays less than twenty items or so, through sheer simplicity and cache friendliness. In other words:

![image](https://user-images.githubusercontent.com/259840/28311264-61fd4610-6baf-11e7-82cf-aa694bc44d6c.png)

I tried various configurations, and the one below seems to beat everything else in both Firefox and Chrome:

```js
// Data structure for a node: 
// [ [array of indices], < char, node, char, node, ... >]
var charCodeArray = [[]];

for (var i = 0; i < words.length; i++) {
  var word = words[i];
  var node = charCodeArray;
  nextchar:
  for (var j = 0; j < word.length; j++) {
    var char = word.charCodeAt(j);
    for (var k = 1; k < node.length; k += 2){
      if (node[k] == char) {
        node = node[k+1];
        node[0].push(i);
        continue nextchar;
      }
    }
    node.push(char);
    var newNode = [[i]];
    node.push(newNode);
    node = newNode;
  }
}
```

No, your eyes are not deceiving you: that's a labelled continue. Yes, I know, we're basically doing low-level programming in JS now.

It's also backwards compatible all the way to IE6 I think :D

Compared to other trie implementations:

|                           | Chrome      | Firefox       |
|---------------------------|-------------|---------------|
| `charCode` array trie     | 76.23±7.22% | 107 ±4.42%    |
| `charCode` trie           | 10.76±9.38% | 32.24 ±7.41%  |
| inverted `charCode` trie  | 22.63±0.53% | 48.43 ±0.62%  |
| `charCode` map            | 63.85±3.29% | 29.07 ±18.19% |
| inverted `charCode` Map   | 48.10±4.59% | 41.22 ±13.93% |
| char-string dictionary    | 21.00±5.59% | 27.79 ±5.82%  |
| char-string map           | 32.33±2.40% | 33.32 ±4.77%  |

And here is how to search it:
```js
function findMatches(prefix, trie, list){
  var node = trie, retList = [];
  nextSearchChar:
  for (var i = 0; i < prefix.length; i++){
    var char = prefix.charCodeAt(i);
    for (var j = 1; j < node.length; j += 2){
      if (node[j] == char){
        node = node[j+1]
        continue nextSearchChar;
      }
    }
    // we only come here if there is
    // no matching char in the trie
    return retList;
  }
  var indices = node[0];
  for (var i = 0; i < indices.length; i++){
    retList.push(list[indices[i]]);
  }
  return retList;
}
```

Try it out here:

https://jsfiddle.net/vanderZwan/ccfxv4sg/1/

I think I might update the LZ-string code to try this out there as well :D

**Comment by JobLeonard** [comment]
As if I didn't spam you enough, here are some more variants to play with :P

This one has an array-based sub-tree for the char codes. By dividing the char code into chunks of four bits, we get 16 element arrays and up to four look-ups. When only dealing with ASCII-codes it performs worse than naive linear search due to extra overhead, but for 16 bit UTF codes it should be much better.

```js
// should have better worst-case behaviour, 
// more memory overhead
// and one-to-four extra look-ups
// (usually two, assuming ASCII codes)
var charCodeArray = [
   null, null, null, null,
   null, null, null, null,
   null, null, null, null,
   null, null, null, null
];

// Data structure for a node: 
// [ <16 nodes for charCode sub-tree>, [<index, index, index...>]]
for (var i = 0; i < words.length; i++) {
   var word = words[i];
   var node = charCodeArray;
   for (var j = 0; j < word.length; j++) {
      var char = word.charCodeAt(j);
      do {
         var charNibble = char & 0xF;
         var nibbleNode = node[charNibble];
         if (!nibbleNode) {
            nibbleNode = [
               null, null, null, null,
               null, null, null, null,
               null, null, null, null,
               null, null, null, null, []
            ];
            node[charNibble] = nibbleNode;
         }
         node = nibbleNode
         char >>>= 4;
      } while (char);
      node[16].push(i);
   }
}
```
https://jsfiddle.net/vanderZwan/jdvsLe3f/

I made a test with 2-, 4- and 8-bit variants, and with and without inlined indices. As usual Chrome and Firefox have different results, although the nibble-version seems to do best on average.

https://jsperf.com/array-based-tries/1

**Comment by bvaughn** [comment]
Just a quick follow-up: I've gotten swamped with work stuff lately so I
haven't had time to dig into your recently replies. They're starred in my
Inbox though for when my schedule clears up a bit. I am looking forward to
it! :)
